bokehjs/src/lib/models/math_text/math_text.ts

```diff
@@ -0,0 +1,226 @@
+
+type Position = {
+  sx: number
+  sy: number
+  x_anchor?: number | "left" | "center" | "right"
+  y_anchor?: number | "top"  | "center" | "baseline" | "bottom"
+}
+
+/**
+ * Helper class to rendering MathText into Canvas
+ */
+export class MathTextView extends View {
+  override model: MathText
+
+  angle?: number
+  font_size_scale: number = 1.0
+  font: string
+  height: number
+  position: Position = {sx: 0, sy: 0}
+  width: number
+  color: string
+  svg_image: CanvasImage
+  image_is_loading = false
+
+  set visuals(v: visuals.Text) {
+    const color = v.text_color.get_value()
+    const alpha = v.text_alpha.get_value()
+    const style = v.text_font_style.get_value()
+    let size = v.text_font_size.get_value()
+    const face = v.text_font.get_value()
+
+    const {font_size_scale} = this
+    if (font_size_scale != 1.0) {
+      const match = size.match(/^\s*(\d+(\.\d+)?)(\w+)\s*$/)
+      if (match != null) {
+        const [, value,, unit] = match
+        const number = Number(value)
+        if (!isNaN(number))
+          size = `${number*font_size_scale}${unit}`
+      }
+    }
+
+    const font = `${style} ${size} ${face}`
+    this.font = font
+    this.color = color2css(color, alpha)
+  }
+
+  /**
+   * Calculates position of element after considering
+   * anchor and dimensions
+   */
+  protected _computed_position(): {x: number, y: number} {
+    const {width, height} = this
+    const {sx, sy, x_anchor="left", y_anchor="center"} = this.position
+
+    const x = sx - (() => {
+      if (isNumber(x_anchor))
+        return x_anchor*width
+      else {
+        switch (x_anchor) {
+          case "left": return 0
+          case "center": return 0.5*width
+          case "right": return width
+        }
+      }
+    })()
+
+    const y = sy - (() => {
+      if (isNumber(y_anchor))
+        return y_anchor*height
+      else {
+        switch (y_anchor) {
+          case "top": return 0
+          case "center": return 0.5*height
+          case "bottom": return height
+          case "baseline": return 0.5*height
+        }
+      }
+    })()
+
+    return {x, y}
+  }
+
+  /**
+   * Uses the width, height and given angle to calculate the size
+  */
+  size(): Size {
+    const {width, height} = this.getDimensions()
+    const {angle} = this
+
+    if (!angle)
+      return {width, height}
+    else {
+      const c = Math.cos(Math.abs(angle))
+      const s = Math.sin(Math.abs(angle))
+
+      return {
+        width: Math.abs(width*c + height*s),
+        height: Math.abs(width*s + height*c),
+      }
+    }
+  }
+
+  /**
+   * Get width and height from SVGElement
+   */
+  private getDimensions(): Size {
+    if (typeof MathJax === "undefined") return {width: 0, height: 0}
+
+    const mathjax_element = MathJax.tex2svg(this.model.text)
+
+    document.body.appendChild(mathjax_element)
+
+    const svg_element = mathjax_element.children[0] as SVGElement
+    svg_element.setAttribute("font", this.font)
```

Comment:
This is still confusing to me. If MathJax can render DOM-less, then why do we need to add the SVG element to the DOM to compute the metrics? I would expect MathJax to to able to provide us with those metrics, or perhaps even those could be read from the SVG content.

Note that adding short lived elements to the DOM tree is not without cost, especially for a pure-canvas rendering engine. This is because to compute `getBoundingClientRect()`, a complete CSS layout has to be computed. Moreover, that happens at the end of the currently evaluating task (in JS run time task queue), which further may degrade performance.

If we can get away without touching the DOM tree, then we should.

Reply:
I've just pushed a version that computes the dimensions without using DOM, I noticed that MathJax returns the width and height of an element in `ex` inside the SVG. I used that and the font_metrics to calculate the dimensions and compared it setting the css of the SVG element and I got the same results.